File: src/domain/conversation/Message.js
/**
 * Message - value object for a single conversation turn.
 *
 * Immutable: create a new Message rather than mutating an existing one.
 */

export const MessageRole = Object.freeze({
  USER: "user",
  ASSISTANT: "assistant",
  UNKNOWN: "unknown",
});

export class Message {
  /**
   * @param {{ role: string, text: string }} params
   */
  constructor({ role, text }) {
    this.role = Object.values(MessageRole).includes(role)
      ? role
      : MessageRole.UNKNOWN;
    this.text = text || "";
    Object.freeze(this);
  }

  isEmpty() {
    return !this.text.trim();
  }

  toDict() {
    return { role: this.role, text: this.text };
  }

  static fromDict(data) {
    return new Message({ role: data.role, text: data.text });
  }
}

File: domain/conversation/message.py
"""Message entity - represents a single turn in a conversation."""
from dataclasses import dataclass
from enum import Enum


class MessageRole(str, Enum):
    """Participant roles in a conversation."""
    USER = "user"
    ASSISTANT = "assistant"
    UNKNOWN = "unknown"


@dataclass
class Message:
    """Represents a single message in a conversation."""
    role: MessageRole
    text: str

    def __post_init__(self):
        if isinstance(self.role, str):
            try:
                self.role = MessageRole(self.role)
            except ValueError:
                self.role = MessageRole.UNKNOWN

    def is_empty(self) -> bool:
        return not self.text.strip()

    def to_dict(self) -> dict:
        return {"role": self.role.value, "text": self.text}

    @classmethod
    def from_dict(cls, data: dict) -> "Message":
        return cls(role=data.get("role", "unknown"), text=data.get("text", ""))
